### PR TITLE
Group by grouping sets

### PR DESCRIPTION
**Pull request by fivetran-dmitryboykov**

### CODE REVIEW

**Comment by pgrivachev** [comment]
@fivetran-dmitryboykov could you please fix commit history here?

File: zetasql/parser/bison_parser.y [hunk 1]
```diff
@@ -5639,6 +5641,17 @@ rollup_list:
       }
     ;
 
+grouping_sets_list:
+    "GROUPING" "SETS" "(" expression
+      {
+        $$ = MAKE_NODE(ASTRollup, @$, {$4});
```
Comment (pgrivachev) [inline zetasql/parser/bison_parser.y]: Why do we have `ASTRollup` here?

File: zetasql/parser/bison_parser.y [hunk 2]
```diff
@@ -5639,6 +5641,17 @@ rollup_list:
...
+      }
+    | grouping_sets_list "," expression
+      {
+        $$ = WithExtraChildren($1, {$3});
```
Comment (pgrivachev) [inline zetasql/parser/bison_parser.y:5651]: Are you sure that we should have `{$3}` here?

File: zetasql/analyzer/resolver_query.cc [hunk 3]
```diff
@@ -443,7 +443,8 @@ absl::Status Resolver::AddAggregateScan(
           column_list, std::move(*current_scan),
           query_resolution_info->release_group_by_columns_to_compute(),
           query_resolution_info->release_aggregate_columns_to_compute(),
-          std::move(grouping_set_list), std::move(rollup_column_list));
+          std::move(grouping_set_list), std::move(rollup_column_list),
+          /*grouping_sets_column_list*/{});
```
Comment (pgrivachev) [inline zetasql/analyzer/resolver_query.cc]: Why `{}`?

File: zetasql/parser/testdata/grouping_sets.test [hunk 4]
```diff
@@ -0,0 +1,259 @@
+SELECT x, SUM(y) FROM T GROUP BY GROUPING SETS(x);
```
Comment (pgrivachev) [inline zetasql/parser/testdata/grouping_sets.test:1]: Can we check that here we have the same result as for 
```
SELECT x, SUM(y) FROM T GROUP BY x;
```
?

File: zetasql/parser/unparser.cc [hunk 5]
```diff
@@ -1546,13 +1546,25 @@ void Unparser::visitASTRollup(const ASTRollup* node, void* data) {
...
 void Unparser::visitASTGroupingItem(const ASTGroupingItem* node, void* data) {
   if (node->expression() != nullptr) {
     ZETASQL_DCHECK(node->rollup() == nullptr);
     node->expression()->Accept(this, data);
   } else {
-    ZETASQL_DCHECK(node->rollup() != nullptr);
-    node->rollup()->Accept(this, data);
+    ZETASQL_DCHECK(node->rollup() != nullptr || node->grouping_sets() != nullptr);
+    if (node->rollup() != nullptr) {
+      ZETASQL_DCHECK(node->rollup() != nullptr);
```
Comment (pgrivachev) [inline zetasql/parser/unparser.cc]: Hm, should we call `ZETASQL_DCHECK` for the condition we checked just before?

File: zetasql/parser/unparser.cc [hunk 6]
```diff
@@ -1546,13 +1546,25 @@ void Unparser::visitASTRollup(const ASTRollup* node, void* data) {
...
+      node->rollup()->Accept(this, data);
+    } else {
+      ZETASQL_DCHECK(node->grouping_sets() != nullptr);
```
Comment (pgrivachev) [inline zetasql/parser/unparser.cc]: The same.

**Review by pgrivachev — APPROVED** [review]

File: zetasql/parser/unparser.cc (hunk 5 above)
Comment (fivetran-dmitryboykov) [reply zetasql/parser/unparser.cc]: Constraint changed

File: zetasql/parser/unparser.cc (hunk 6 above)
Comment (fivetran-dmitryboykov) [reply zetasql/parser/unparser.cc]: Constraint changed

File: zetasql/parser/bison_parser.y (hunk 1 above)
Comment (fivetran-dmitryboykov) [reply zetasql/parser/bison_parser.y]: Not relevant anymore.

File: zetasql/parser/bison_parser.y (hunk 2 above)
Comment (fivetran-dmitryboykov) [reply zetasql/parser/bison_parser.y:5651]: 'WithExtraChildren' adds 'children' to 'node' and then returns 'node'.

File: zetasql/analyzer/resolver_query.cc (hunk 3 above)
Comment (fivetran-dmitryboykov) [reply zetasql/analyzer/resolver_query.cc]: Not relevant anymore.

File: zetasql/parser/testdata/grouping_sets.test (hunk 4 above)
Comment (fivetran-dmitryboykov) [reply zetasql/parser/testdata/grouping_sets.test:1]: Parse tree with `GROUP BY` differs from parse tree with `GROUP BY GROUPING SETS`.

**Pull request merged**